File: go/internal/commands/set_state.go
package commands

import (
	"fmt"
	"strings"

	"all-you-need-is-git/go/internal/statex"
)

type SetStateOptions struct {
	State       string
	Subject     string
	Prompt      string
	PromptFile  string
	PromptStdin bool
	DwpRemote   string
	Trailers    []string
}

func SetState(opts SetStateOptions) error {
	state := strings.ToLower(strings.TrimSpace(opts.State))
	if state == "" {
		return fmt.Errorf("Missing required flag: --dwp-state")
	}
	if state == "working" {
		return fmt.Errorf("Invalid dwp-state: working (use aynig set-working)")
	}

	fullMessage, err := gitShowHeadBody()
	if err != nil {
		return err
	}
	_, headBody := splitCommitMessage(fullMessage)
	headTrailers, err := parseTrailersFromBody(headBody)
	if err != nil {
		return err
	}

	remote := resolveDwpRemote(opts.DwpRemote, headTrailers)
	if err := validateRemoteExists(remote); err != nil {
		return err
	}

	prompt, err := readPrompt(promptOptions{Prompt: opts.Prompt, PromptFile: opts.PromptFile, PromptStdin: opts.PromptStdin}, "")
	if err != nil {
		return err
	}

	subject := strings.TrimSpace(opts.Subject)
	if subject == "" {
		subject = fmt.Sprintf("chore: set %s", state)
	}

	trailers := []statex.Trailer{{Key: "dwp-state", Value: state}}
	if remote != "" {
		trailers = append(trailers, statex.Trailer{Key: "dwp-source", Value: "git:" + remote})
	}

	for _, raw := range opts.Trailers {
		parsed, parseErr := parseTrailerArg(raw)
		if parseErr != nil {
			return parseErr
		}
		if strings.EqualFold(strings.TrimSpace(parsed.Key), "dwp-state") {
			return fmt.Errorf("Invalid trailer: dwp-state is managed by --dwp-state")
		}
		trailers = append(trailers, parsed)
	}

	if err := statex.CommitState("", subject, prompt, trailers); err != nil {
		return err
	}

	return pushCurrentBranch(remote)
}

File: js/commands/set-state.js
import { commitState } from '../AgentsOrchestrator/stateCommit.js';
import { git } from '../gitHelpers/git.js';
import {
  parseTrailerArg,
  pushCurrentBranch,
  readHeadCommitMessage,
  readPrompt,
  resolveAynigRemote,
  validateRemoteExists
} from './stateCommon.js';

export async function action(options) {
  const state = String(options.aynigState || '').trim().toLowerCase();
  if (!state) {
    throw new Error('Missing required flag: --aynig-state');
  }
  if (state === 'working') {
    throw new Error('Invalid aynig-state: working (use aynig set-working)');
  }

  const parsed = await readHeadCommitMessage();
  const headTrailers = parsed.trailers || {};

  const remote = resolveAynigRemote(options.aynigRemote, headTrailers);
  await validateRemoteExists(remote);

  const prompt = await readPrompt(options, '');
  const subject = (options.subject || '').trim() || `chore: set ${state}`;

  const trailers = [{ key: 'aynig-state', value: state }];
  if (remote) {
    trailers.push({ key: 'aynig-remote', value: remote });
  }

  for (const raw of options.trailer || []) {
    const parsedTrailer = parseTrailerArg(raw);
    if (parsedTrailer.key.trim().toLowerCase() === 'aynig-state') {
      throw new Error('Invalid trailer: aynig-state is managed by --aynig-state');
    }
    trailers.push(parsedTrailer);
  }

  await commitState(git, subject, prompt, trailers);
  await pushCurrentBranch(remote);
}

export function registerSetStateCommand(program) {
  program
    .command('set-state')
    .description('Create a commit with a new non-working aynig-state')
    .requiredOption('--aynig-state <state>', 'Next state (must not be working)')
    .option('--subject <text>', 'Commit title')
    .option('--prompt <text>', 'Commit prompt/body')
    .option('--prompt-file <path>', 'Path to file used as prompt/body')
    .option('--prompt-stdin', 'Read prompt/body from stdin')
    .option('--aynig-remote <name>', 'Remote name to push after commit')
    .option('--trailer <key:value>', 'Additional trailer (repeatable)', (value, acc) => {
      acc.push(value);
      return acc;
    }, [])
    .action(async (options) => {
      try {
        await action(options);
      } catch (error) {
        console.error(error.message || error);
        process.exit(1);
      }
    });
}
